### PR TITLE
Added mixture weight to combine entity and action logits.

### PR DESCRIPTION
**Pull request by rajasagashe**

### CODE REVIEW

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py [hunk 1]
```diff
@@ -1063,6 +1069,9 @@ def __init__(self,
         # TODO(pradeep): Do not hardcode decoder cell type.
         self._decoder_cell = LSTMCell(input_dim, output_dim)
 
+        check_dimensions_match(output_dim, mixture_feedforward.get_input_dim(),
```
Comment (matt-gardner) [inline allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: You should probably also check that the output dimension of the mixture feedforward is 1.

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py [hunk 2]
```diff
@@ -1117,6 +1126,14 @@ def take_step(self,
             # entity_action_mask: (group_size, num_entity_actions)
             entity_action_logits, entity_action_mask, entity_type_embeddings = \
                     self._get_entity_action_logits(state, actions_to_link, attention_weights)
+
+            # The entity and action logits are combined with a mixture weight to prevent the
```
Comment (matt-gardner) [inline allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: I'd put this behind a flag - allow the `mixture_feedforward_params` to be `None`, and if they are, you skip the mixture.  That will let us try both versions side-by-side with just a configuration change, instead of a code change.

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py [hunk 3]
```diff
@@ -1117,6 +1126,14 @@ def take_step(self,
...
+            # entity_action_logits from dominating the embedded_action_logits if a softmax
+            # was applied on both together.
+            mixture_weight = self._mixture_feedforward(hidden_state)
+            entity_action_logits = util.masked_softmax(entity_action_logits, entity_action_mask) * mixture_weight
+            embedded_action_logits = util.masked_softmax(embedded_action_logits, embedded_action_mask) * mixture_weight
```
Comment (matt-gardner) [inline allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: This is close, but not quite right, because these aren't logits anymore, after you've done a softmax.  These are now probabilities.  So, you can't do another softmax with these below - some more logic needs to be modified to make this work right.

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py (hunk 1 above)
Comment (rajasagashe) [reply allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: done.

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py (hunk 2 above)
Comment (rajasagashe) [reply allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: done.

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py (hunk 3 above)
Comment (rajasagashe) [reply allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: done.

**Comment by matt-gardner** [comment]
I rebased the wikitables branch this morning, so you'll need to rebase your PRs, too.  Hopefully soon we can have this working well enough that we can just merge the model into master, and won't have to do all of this rebasing...

**Comment by rajasagashe** [comment]
@matt-gardner  there's way too many conflicts after the rebase. Should i create a new branch from your rebased wikitables and cherry pick these two commits on top of that?

**Comment by rajasagashe** [comment]
Thanks the interactive flag makes it a lot easier. 
The hidden states that are passed to the mixture_feedforward have dimensions (num_hidden_states,  hidden_dim), meaning that the mixture_weight will have dimensions (num_hidden_states, 1). Any ideas on how to make sure mixture_weight has dimensions (1, 1)?

**Comment by matt-gardner** [comment]
I'm not sure I understand your question - your check on the output dimension of the mixture feedforward should already accomplish that, right?

**Comment by rajasagashe** [comment]
Lets say the hidden_states variable is (num_hidden_states, hidden_state_dim). Then after it goes through the feedforward it's (num_hidden_states, 1). So the check for output dimension 1 does make sure the second dimension is 1.  I'm trying to figure out how to collapse the num_hidden_states to 1.

**Comment by rajasagashe** [comment]
Actually never mind, when num_hidden_states = 3, the first dimension of entity_action_logits and embedded_action_logits is also 3. So outputting a (num_hidden_states, 1) vector is ok.

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py [hunk 4]
```diff
@@ -1127,13 +1143,38 @@ def take_step(self,
...
+            if self._mixture_feedforward is not None:
+                # The entity and action logits are combined with a mixture weight to prevent the
+                # entity_action_logits from dominating the embedded_action_logits if a softmax
+                # was applied on both together.
+                # Todo(rajas) remove mix1 and mix2 variables
+                # Todo(rajas) figure out why the loss is nan with this code.
+                mixture_weight = self._mixture_feedforward(hidden_state)
+                mix1 = (torch.log(mixture_weight) * entity_action_mask.float())
+                mix2 = (torch.log(1 - mixture_weight) * embedded_action_mask.float())
+                entity_action_logits = util.masked_log_softmax(entity_action_logits, entity_action_mask.float()) + mix1
```
Comment (matt-gardner) [inline allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: This looks right, I'm not sure why you would be getting nans here.  Maybe some value is just initialized too high, and so training is unstable?  You could try printing out the mixture weight and the values of mix1 and mix2, to see if something looks wrong.

Also, because you've done a log softmax, these are no longer logits, they are log probabilities.

**Comment by rajasagashe** [comment]
I fixed the nan issue by changing the config file for the feedforward from:
```
"mixture_feedforward": {
        "input_dim": 200,
        "num_layers": 3,
        "hidden_dims": [25, 1],
        "activations": ["relu", "sigmoid"],
        "dropout": [0.2, 0.2]
    },
```
to 
```
"mixture_feedforward": {
        "input_dim": 200,
        "num_layers": 3,
        "hidden_dims": [25, 2, 1],
        "activations": ["relu", "sigmoid", "sigmoid"],
        "dropout": [0.2, 0.0, 0.0]
    },
```
Removing the dropout reduced the nans a lot, and adding another sigmoid layer stopped them entirely.
This improves the parser's validation baseline accuracy by 2.5%

**Comment by rajasagashe** [comment]
@matt-gardner  I think it's ready now.

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py [hunk 5]
```diff
@@ -129,7 +130,8 @@ def __init__(self,
         self._decoder_step = WikiTablesDecoderStep(encoder_output_dim=self._encoder.get_output_dim(),
                                                    action_embedding_dim=action_embedding_dim,
                                                    attention_function=attention_function,
-                                                   num_entity_types=num_entity_types)
+                                                   num_entity_types=num_entity_types,
+                                                   mixture_feedforward=mixture_feedforward,)
```
[the remark this answers is not in the document]
Comment (rajasagashe) [reply allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: done

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py [hunk 6]
```diff
@@ -1127,18 +1143,47 @@ def take_step(self,
...
+
+            action_logits = torch.cat([embedded_action_logits, entity_action_logits], dim=1)
+            action_mask = torch.cat([embedded_action_mask, entity_action_mask], dim=1).float()
+
         else:
             action_logits = embedded_action_logits
             action_mask = embedded_action_mask.float()
         log_probs = util.masked_log_softmax(action_logits, action_mask)
-
+        # print('---------------')
```
[the remark this answers is not in the document]
Comment (rajasagashe) [reply allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: done

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py [hunk 7]
```diff
@@ -1127,18 +1143,47 @@ def take_step(self,
...
+                mix1 = torch.log(mixture_weight)
+                mix2 = torch.log(1 - mixture_weight)
+
+                entity_action_probs = util.masked_log_softmax(entity_action_logits,
+                                                              entity_action_mask.float()) + mix1
+                embedded_action_probs = util.masked_log_softmax(embedded_action_logits,
+                                                                embedded_action_mask.float()) + mix2
+                log_probs = torch.cat([embedded_action_probs, entity_action_probs], dim=1)
+
+                return self._compute_new_states(state,
```
[the remark this answers is not in the document]
Comment (rajasagashe) [reply allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: done

**Comment by rajasagashe** [comment]
Also, the test I added will pass once the fix to masked_log_softmax is integrated into this branch.

**Comment by matt-gardner** [comment]
I just rebased the wikitables branch onto the current master, so you should be able to rebase your PR and get the test to pass now.

File: tests/fixtures/encoder_decoder/wikitables_semantic_parser_with_mixture_feedforward/serialization/vocabulary/non_padded_namespaces.txt [hunk 8]
```diff
@@ -0,0 +1,2 @@
+*tags
```
Comment (matt-gardner) [inline tests/fixtures/encoder_decoder/wikitables_semantic_parser_with_mixture_feedforward/serialization/vocabulary/non_padded_namespaces.txt]: You shouldn't need to add a `serialization` directory here - you only need that to test `Predictors`.  I think these fixtures are just large unused files.

Comment (rajasagashe) [reply tests/fixtures/encoder_decoder/wikitables_semantic_parser_with_mixture_feedforward/serialization/vocabulary/non_padded_namespaces.txt]: done.

File: tests/models/encoder_decoders/wikitables_semantic_parser_test.py [hunk 9]
```diff
@@ -305,6 +305,15 @@ def test_map_entity_productions(self):
                 }
 
 
+class WikiTablesSemanticParserTestWithMixtureFeedforward(ModelTestCase):
+    def setUp(self):
+        super(WikiTablesSemanticParserTestWithMixtureFeedforward, self).setUp()
+        self.set_up_model("tests/fixtures/encoder_decoder/wikitables_semantic_parser_with_mixture_feedforward/experiment.json", # pylint: disable=line-too-long
```
Comment (matt-gardner) [inline tests/models/encoder_decoders/wikitables_semantic_parser_test.py]: Ok, last nitpick, then I think this is good to merge.  Can you move this line inside the `test_model...` method, then move that method inside the already-existing class?  There isn't really a reason to create a new class for this.

Great work on this!

**Comment by matt-gardner** [comment]
That test failure isn't due to the iterator, it's due to something else.  See this commit: https://github.com/allenai/allennlp/pull/936/commits/35d58ef081163981d19baff147725bdcd3850811.

**Comment by rajasagashe** [comment]
Isn't that commit in the wikitables branch?

**Comment by matt-gardner** [comment]
It's not yet, it's in #915.

**Comment by rajasagashe** [comment]
I think this is ready to merge.

**Comment by matt-gardner** [comment]
Yep, sorry I forgot about it!
